---
title: "Compositional Regression"
date: "`r Sys.Date()`"
runtime: html
output: 
  html_document:
    fig_width: 16
    fig_height: 10
    fig_caption: true
    number_sections: true
    toc: false
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, dev = 'svg')

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####
rds_path <- "/Users/reuben/Library/CloudStorage/OneDrive-SharedLibraries-CranfieldUniversity/Sophie Bretagne - Shared/data/experiment_genus.rds"

grouping_column <- "Groups"
number_of_randomizations <- 1000
distance_measure <- "cao"
abundance.weighted <- FALSE
null_model <- "PF"
SES <- TRUE
RC <- FALSE
```

```{r data_import, include=FALSE, echo=TRUE}
library(phyloseq)

#### Data Import ####
physeq <- import_rds(rds_path)

# Here we choose what taxonomic rank to cluster around
# physeq <- collate_taxonomy_2(physeq, rank = "Genus")

abund_table <- otu_table(physeq, TRUE)
taxa_table <- tax_table(physeq, TRUE)
meta_table <- sample_data(physeq, TRUE)
taxa_tree <- phy_tree(physeq, FALSE)

# We can proceed without a phylogenetic tree but some analyses will be unavailable
if (is.null(taxa_tree)) {
  warning("No phylogenetic tree was imported. Proceeding, but some analyses may be unavailable.")
}
```

```{r data_check, include=FALSE}
stopifnot({
  exists("abund_table")
  exists("taxa_table")
  exists("meta_table")
  exists("taxa_tree")
})
```

```{r hypothesis, include=FALSE}
meta_table$Groups <- as.factor("All")

explanatory_variables <- c(
  "pH",
  "Headloss",
  "temperature..C.",
  "Neat.coliform.2000",
  "neat.e.coli.2000",
  "Cl.perfrigens.conf",
  "Enterococci.confirmed.100ml",
  "X22C.Plate.Count.Neat",
  "Conductivity.20C",
  "Turbidity.FTU",
  "Ammoniacal.Nitrogen",
  "Ammonium.NH4",
  "Nitrate.as.N",
  "Nitrite.as.N",
  "Alkalinity.as.CaCO3",
  "Sulphate.as.SO4",
  "Carbon.Total.Organic"
)
```

```{r adjust, include=FALSE}
if (taxa_are_rows(abund_table)) {
  abund_table <- t(abund_table)
}
abund_table <- as.data.frame(abund_table)

#Adjust abund_table to contain only those rows that got selected in the Hypothesis space
abund_table <- abund_table[rownames(meta_table),]
#After adjustment, get rid of OTUs that are all empty
abund_table <- abund_table[,colSums(abund_table) > 0]
#Adjust OTU taxonomy
taxa_table <- taxa_table[colnames(abund_table),]
```

# Compositional Regression

Due to the compositional nature of microbiome data, we use a balance approach to find a minimal subset of genera whose abundance correlates with each covariate. 

CODA LASSO regression using the coda_glmnet() function from Râ€™s coda4microbiome package was applied for the top 100 most abundant genera (Calle et al., 2023). The fitted regression was of the form y_i=\beta_0+\beta_1\log{(x_{1i})}+\ldots+\beta_j\log{\left(x_{ji}\right)}+\epsilon_i (for ð‘–-th sample and ð‘—-th feature, with x_{ji} being the abundance of feature), and where the outcome y_i was a continuous outcome variable. 

The model used two constraints: a) all ð›½-coefficients summed up to zero, that gave two sets of features: those that were positively associated with continuous outcome, and those that were otherwise; and b) the optimization function incorporated a LASSO shrinkage that forced some of the beta coefficients to go to zero, particularly those that were statistically insignificant in relation to the outcome.

```{r analyse, include=FALSE}
new_physeq <- join_phyloseq(abund_table, taxa_table, meta_table, taxa_tree)

result <- coda_glmnet(new_physeq, explanatory_variables, normalisation_method = "TSS+CLR")
```

```{r plot, echo=FALSE}
p <- plot.ECCompositionalRegression(result) + theme_light() + theme(
  strip.background = element_rect(fill = "#cdcdcd"),
  strip.text = element_text(color = "black")
)

print(p)
```
