---
title: "Core Taxa"
date: "`r Sys.Date()`"
#runtime: shiny
output: 
  html_document:
    fig_width: 16
    fig_height: 10
    fig_caption: true
    number_sections: true
    toc: false
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE)

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####
#biom_path="/Users/reuben/test_data/data/Batch 1/asv_even_taxon.biom"
#meta_path="/Users/reuben/test_data/data/metadata.csv"
#tree_path="/Users/reuben/test_data/data/Batch 1/tree_rooted.nwk"
#grouping_column<-"Groups"

biom_path="data/DADA2/feature_w_tax.biom"
meta_path="data/meta_table.csv"
tree_path="data/DADA2/tree.nwk"
```

```{r import_data, include=FALSE}
#### Data Import ####
library(phyloseq)
#experiment <- import_data(biom_path, meta_path, tree_path)
experiment <- import_data(biom_path, meta_path)
#abund_table <- result[[1]]
#taxa_table <- result[[2]]
#meta_table <- result[[3]]
#taxa_tree <- result[[4]]

#rm(result)
```
```{r summary}
read_counts <- sample_sums(experiment)

summary(read_counts)
```


```{r qpcr}
# Later, we're going to fit a neutral model to our core microbiome in order to see which members are selected for deterministically
# The Sloan neutral model we're using assumes a constant coomunity size (N), but library sizes can vary by orders of magnitude between samples
# So, we have two options: either rarefy the data, or in this case we have total microbial count data in the form of 16S qPCR results, which we can use to give us a mean value for microbial counts

ta <- as.data.frame(otu_table(experiment))
ma <- metadata_frame(experiment)

library_size <- as.data.frame(rowSums(ta))

ta_scaled <- as.matrix(ta) * as.numeric(as.character(ma[row.names(ta), "qPCR_16S_per_ml"])) / as.numeric(as.character(library_size[row.names(library_size), "rowSums(ta)"]))

ta_scaled_f <- t(ta_scaled)

total_reads <- colSums(ta_scaled_f)

total_reads_df <- data.frame(Sample = names(total_reads), TotalReads = total_reads)

p <- ggplot(total_reads_df, aes(x = TotalReads)) + geom_density()

print(p)

mean_N <- mean(total_reads)
```

# Core Microbiome

The core microbiome captures the set of microbial taxa consistently found across a given group of samples.

https://www.pnas.org/doi/full/10.1073/pnas.2104429118#body-ref-r48

## Abundance-Occupancy Model

```{r core, include=FALSE}
core <- core_occupancy_abundance_model(experiment)
```

```{r core_oa_plot, echo=FALSE}
plot.ECCOccupancyAbundance(core)
```

```{r core_rs_plot, echo=FALSE}
plot.ECCRankedSimilarity(core)
```

These species play an active role in shaping the ecological community.

```{r core_rs_plot, echo=FALSE}
p <- plot.ECCHeatmap(core)



print(p)
```

## Abundance

```{r core_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Abstraction_Method"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

meta_table$Type <- NULL
meta_table$Type2 <- NULL
meta_table$Connections <- NULL
```

```{r collate, include=FALSE}
#Adjust abund_table to contain only those rows that got selected in the Hypothesis space
abund_table<-abund_table[rownames(meta_table),]
#After adjustment, get rid of OTUs that are all empty
abund_table<-abund_table[,colSums(abund_table)>0]
#Adjust OTU taxonomy
taxa_table<-taxa_table[colnames(abund_table),]
```

```{r core_microbiome, echo=FALSE}
# Loop through and create a core plot for each unique group
plots <- list()
for (group in unique(meta_table$Groups)) {
  meta_sub <- meta_table[meta_table$Groups == group,]
  
  data <- core_microbiome(abund_table, meta_sub, taxa_table, taxa_tree, what_detection = "absolute", short_names = TRUE)
  
  plots <- append(plots, list(data))
}

print(do.call(gridExtra::grid.arrange, plots))
```

## Taxa Bars

```{r core_taxa, echo=FALSE}
data <- core_taxa(abund_table, meta_table, taxa_table, taxa_tree)
plot <- plot_core_taxa(data)

print(plot)
```

## Abundance-Occupancy Models

## Hypothesis Setting

## Plots
