---
title: "Differential Abundance"
date: "`r Sys.Date()`"
runtime: html
output: 
  html_document:
    fig_width: 16
    fig_height: 10
    fig_caption: true
    number_sections: true
    toc: false
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, dev = 'svg')

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####
rds_path <- "/Users/reuben/Library/CloudStorage/OneDrive-SharedLibraries-CranfieldUniversity/Sophie Bretagne - Shared/data/experiment_genus.rds"

which_level<-"Otus" #Phylum Class Order Family Genus Otus
sig = 0.05
fold = 2
height_image=15
```

```{r data_import, include=FALSE, echo=TRUE}
library(phyloseq)

#### Data Import ####
physeq <- import_rds(rds_path)

# Here we choose what taxonomic rank to cluster around
# physeq <- collate_taxonomy_2(physeq, rank = "Genus")

abund_table <- otu_table(physeq, TRUE)
taxa_table <- tax_table(physeq, TRUE)
meta_table <- sample_data(physeq, TRUE)
taxa_tree <- phy_tree(physeq, FALSE)

# We can proceed without a phylogenetic tree but some analyses will be unavailable
if (is.null(taxa_tree)) {
  warning("No phylogenetic tree was imported. Proceeding, but some analyses may be unavailable.")
}
```

```{r data_check, include=FALSE}
stopifnot({
  exists("abund_table")
  exists("taxa_table")
  exists("meta_table")
  exists("taxa_tree")
})
```

```{r hypothesis, include=FALSE}
meta_table$Groups<-as.factor(as.character(meta_table$run))

#Hypothesis 1
labels="Run 1"
meta_table<-meta_table[meta_table$Groups %in% c("Run 1"),]

#Hypothesis 2
labels="Run 2"
meta_table<-meta_table[meta_table$Groups %in% c("Run 2"),]

#Hypothesis 3
labels="Run 3"
meta_table<-meta_table[meta_table$Groups %in% c("Run 3"),]

#Hypothesis 4
labels="Run 4"
meta_table<-meta_table[meta_table$Groups %in% c("Run 4"),]

#Hypothesis 5
labels="Run 5"
meta_table<-meta_table[meta_table$Groups %in% c("Run 5"),]

#Hypothesis 6
labels="Run 6"
meta_table<-meta_table[meta_table$Groups %in% c("Run 6"),]

#Hypothesis 7
labels="Run 7"
meta_table<-meta_table[meta_table$Groups %in% c("Run 7"),]

#Hypothesis 8
labels="Run 8"
meta_table<-meta_table[meta_table$Groups %in% c("Run 8"),]
```

```{r adjust, include=FALSE}
if (taxa_are_rows(abund_table)) {
  abund_table <- t(abund_table)
}
abund_table <- as.data.frame(abund_table)

#Adjust abund_table to contain only those rows that got selected in the Hypothesis space
abund_table <- abund_table[rownames(meta_table),]
#After adjustment, get rid of OTUs that are all empty
abund_table <- abund_table[,colSums(abund_table) > 0]
#Adjust OTU taxonomy
taxa_table <- taxa_table[colnames(abund_table),]
```

# Differential Abundance

If the core determines which taxa are in the intersection of a sample set, differential expression analyses the difference between sample sets. It is used to determine which taxa exhibit statistically significant differences in abundance between pairs of groups.

DESeq2 is a statistical package originally developed for analysing count data from RNA-Seq experiments but has been adapted for microbial count data. It uses a model based on the negative binomial distribution to normalise count data and identify differentially abundant taxa, accounting for biological and technical variation.

In microbial ecology, DESeq2 can highlight taxa whose relative abundances change under different environmental or experimental conditions, thus revealing potential biomarkers or organisms that drive community shifts.

Other models to consider include the following: 

Bulk RNA-Seq:
- DESeq2
- edgeR
- limma-voom

No replicates:
- NOISeq
- DEGseq

Single-Cell RNA-Seq:
- MAST
- Seurat
- Monocle3
- glmGamPoi

Bayesian Approaches:
- EBSeq
- baySeq
- BPSC
- ABSSeq

See this paper for more: <https://www.nature.com/articles/nmeth.2658>

```{r analyse, include=FALSE}
result <- de_create()
```

```{r plot, echo=FALSE}
de_summarise_results()

de_get_results()

plot_differential_expression()
```
