---
title: "Differential Expression"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE)

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####

biom_path <- "/Users/reuben/Dropbox/Cindy_Alison_Umer_SandDam/data/DADA2/feature_w_tax.biom"
meta_path <- "/Users/reuben/Dropbox/Cindy_Alison_Umer_SandDam/data/meta_table.csv"
tree_path <- "/Users/reuben/Dropbox/Cindy_Alison_Umer_SandDam/data/DADA2/tree.nwk"
grouping_column <- "Groups"

which_level<-"Otus" #Phylum Class Order Family Genus Otus
sig = 0.05
fold = 2
height_image=15

```

## Core Microbiome

This notebook uses the EcologyCore API to generate core microbiome plots from a .biom file.

```{r data, include=FALSE}
#### Data Import ####

result <- import_data(biom_path, meta_path, tree_path)

abund_table <- result[[1]]
OTU_taxonomy <- result[[2]]
meta_table <- result[[3]]
OTU_tree <- result[[4]]

rm(result)
```

## Setting up Hypotheses

The hypothesis space sets up comparison groups for the analysis.

All you need to do is select the metadata columns you are interested in comparing and the code groups samples by the values in these columns.

Additionally, you can allocate a column to Connections if the data is paired.

```{r hypothesis, include=FALSE}
#### Hypothesis ####

meta_table<-meta_table[!meta_table$Within_Dam_Sample %in% c("Piezometer"),]
meta_table$Groups<-paste(meta_table$Within_Dam_Sample,meta_table$Sample_Time)


#Hypothesis 1
labels="Excavated_scoop_hole"
meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  May","Excavated Scoop Hole  July"),]
meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  May","Excavated Scoop Hole  July"))

# #Hypothesis 2
# labels="Community_scoop_hole"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole May","Community Scoop Hole July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole May","Community Scoop Hole July"))

# #Hypothesis 3
# labels="Hand_pump"
# meta_table<-meta_table[meta_table$Groups %in% c("Hand Pump May","Hand Pump July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Hand Pump May","Hand Pump July"))

# #Hypothesis 4
# labels="Open_well"
# meta_table<-meta_table[meta_table$Groups %in% c("Open Well May","Open Well July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Open Well May","Open Well July"))

# #Hypothesis 5
# labels="Downstream"
# meta_table<-meta_table[meta_table$Groups %in% c("Downstream May","Downstream July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Downstream May","Downstream July"))

# #Hypothesis 6
# labels="Excavated Scoop Hole  May_Community Scoop Hole May"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  May","Community Scoop Hole May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  May","Community Scoop Hole May"))

# #Hypothesis 7
# labels="Excavated Scoop Hole  May_Open Well May"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  May","Open Well May"  ),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  May","Open Well May"  ))

# #Hypothesis 8
# labels="Excavated Scoop Hole  May_Hand Pump May"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  May","Hand Pump May" ),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  May","Hand Pump May" ))

# #Hypothesis 9
# labels="Excavated Scoop Hole  May_Downstream May"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  May","Downstream May" ),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  May","Downstream May" ))

# #Hypothesis 10
# labels="Community Scoop Hole May_Open Well May"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole May","Open Well May" ),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole May","Open Well May" ))

# #Hypothesis 11
# labels="Community Scoop Hole May_Hand Pump May"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole May","Hand Pump May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole May","Hand Pump May"))

# #Hypothesis 12
# labels="Community Scoop Hole May_Downstream May"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole May","Downstream May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole May","Downstream May") )

# #Hypothesis 13
# labels="Open Well May_Hand Pump May"
# meta_table<-meta_table[meta_table$Groups %in% c("Open Well May","Hand Pump May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Open Well May","Hand Pump May"))

# #Hypothesis 14
# labels="Open Well May_Downstream May"
# meta_table<-meta_table[meta_table$Groups %in% c("Open Well May","Downstream May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Open Well May","Downstream May"))

# #Hypothesis 15
# labels="Hand Pump May_Downstream May"
# meta_table<-meta_table[meta_table$Groups %in% c("Hand Pump May","Downstream May"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Hand Pump May","Downstream May"))

# #Hypothesis 16
# labels="Community Scoop Hole July_Excavated Scoop Hole  July"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole July","Excavated Scoop Hole  July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole July","Excavated Scoop Hole  July"))

# #Hypothesis 17
# labels="Community Scoop Hole July_Hand Pump July"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole July","Hand Pump July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole July","Hand Pump July"))

# #Hypothesis 18
# labels="Community Scoop Hole July_Open Well July"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole July","Open Well July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole July","Open Well July"))

# #Hypothesis 19
# labels="Community Scoop Hole July_Downstream July"
# meta_table<-meta_table[meta_table$Groups %in% c("Community Scoop Hole July","Downstream July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Community Scoop Hole July","Downstream July"))

# #Hypothesis 20
# labels="Excavated Scoop Hole  July_Hand Pump July"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  July", "Hand Pump July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  July", "Hand Pump July"))

# #Hypothesis 21
# labels="Excavated Scoop Hole  July_Open Well July"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  July","Open Well July" ),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  July","Open Well July" ))

# #Hypothesis 22
# labels="Excavated Scoop Hole  July_Downstream July"
# meta_table<-meta_table[meta_table$Groups %in% c("Excavated Scoop Hole  July", "Downstream July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Excavated Scoop Hole  July", "Downstream July"))

# #Hypothesis 23
# labels="Hand Pump July_Open Well July"
# meta_table<-meta_table[meta_table$Groups %in% c("Hand Pump July","Open Well July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Hand Pump July","Open Well July"))

# #Hypothesis 24
# labels="Hand Pump July_Downstream July"
# meta_table<-meta_table[meta_table$Groups %in% c("Hand Pump July","Downstream July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Hand Pump July","Downstream July"))

# #Hypothesis 25
# labels="Open Well July_Downstream July"
# meta_table<-meta_table[meta_table$Groups %in% c("Open Well July","Downstream July"),]
# meta_table$Groups<-factor(meta_table$Groups,levels=c("Open Well July","Downstream July") )



#Adjust abund_table to contain only those rows that got selected in the Hypothesis space
abund_table<-abund_table[rownames(meta_table),]
#After adjustment, get rid of OTUs that are all empty
abund_table<-abund_table[,colSums(abund_table)>0]
#Adjust OTU taxonomy
OTU_taxonomy<-OTU_taxonomy[colnames(abund_table),]
```

## Collate Taxonomy

This function is included in every analysis and serves to coalesce values in the abundance table at whatever taxonomic rank you specify.

```{r collate, include=FALSE}
#### Collate Taxonomy ####

abund_table <- collate_taxonomy(abund_table, OTU_taxonomy, which_level)
```

## Model Selection

There are a wide variety of models available for analysing differential expression.

Bulk RNA-Seq:
- DESeq2
- edgeR
- limma-voom

No replicates:
- NOISeq
- DEGseq

Single-Cell RNA-Seq:
- MAST
- Seurat
- Monocle3
-glmGamPoi

Bayesian Approaches:
- EBSeq
- baySeq
- BPSC
- ABSSeq

```{r analyse, include=FALSE}
#### Analysis ####

dt <- alpha_diversity(abund_table, meta_table, grouping_column)

# Some absolutely fucking insufferable bug causes this to fail if its run in a package
# so fuck it I'm sticking it here instead
pval<-data.table::as.data.table(dt)[, list(pvalue = sprintf("%.2g", tryCatch({
   summary(stats::aov(value ~ .SD$.group.))[[1]][["Pr(>F)"]][1]
  }, error = function(e) NULL))),
  by = list(measure)
]
  
result <- alpha_diversity_2(dt, pval, meta_table, grouping_column)

df <- result[[1]]
df_pw <- result[[2]]

rm(result)
```

## Including Plots

Once the analysis is complete, a separate function call draws the plots, and these are rendered here in the notebook.

```{r plot, echo=FALSE}
plot_alpha_diversity(df, df_pw)
```

## Saving Data

It is good practice to save the data produced by the analyses in a non-destructive way. This means that someone coming back to the analysis after some time can see the output of each step, for validation and reproducibility. This also saves one running the entire analysis again in order to redraw a plot or change one final piece. This becomes particularly useful for large, complex analyses that may take hours or even days to complete.

By default, we save a workspace image containing the output of each analysis step in an RData file. This provides an easy way for the user to open the data created by this software and analyse it further, independently. For users that don't use R, or that want a more portable format, we provide an option to export data in CSV format.

```{r save, include=FALSE}
save_plot(plot, paste("NB_significant_",which_level,"_",labels,".pdf",sep=""), width = ceiling((length(sig_otus) * 80 / 200) + 2.6), height = height_image)
```
