---
title: "Full Workflow"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_width: 16
    fig_height: 10
    fig_caption: true
    number_sections: true
    toc: true
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE)

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####

biom_path="/Users/reuben/test_data/data/Batch 1/asv_even_taxon.biom"
meta_path="/Users/reuben/test_data/data/metadata.csv"
tree_path="/Users/reuben/test_data/data/Batch 1/tree_rooted.nwk"
grouping_column<-"Groups"

which_level="Otus" #Otus Genus Family Order Class Phylum
taxa_rank <- which_level
```

# Introduction

This workflow runs through a number of ecological analyses in series as follows:

- **Alpha Diversity:** Richness, Shannon Index, Simpson Index, Fisher's Alpha, and Pielou's Evenness
- **Beta Diversity:** Bray-Curtis, Unweighted UniFrac, and Weighted UniFrac
- **Core Microbiome:** taxa bars and core heatmaps combine to show the most abundant species in each sample + the most prevalent in each group
- **Differential Expression:** DESeq2; shows significant differences in abundance between pairs of groups
- **Community Assembly Processes:** Normalised Stochasticity Ratio, Quantitative Process Estimate
- **Environmental Filtering:** NRI, NTI; environmental impact on evolution of phylogeny
- **Subset Regression:** regress metadata against diversity metrics (alpha, NRI, NTI, LCBD, etc.)
- **Variable Selection:** Selbal, CLR LASSO, CODA LASSO; selects two groups of taxa whose balance is most associated with the response (metadata) variable

In combination, these analyses are intended to give a complete picture of an ecological system.

```{r import_data, include=FALSE}
#### Data Import ####

result <- import_data(biom_path, meta_path, tree_path)

abund_table <- result[[1]]
taxa_table <- result[[2]]
meta_table <- result[[3]]
taxa_tree <- result[[4]]

rm(result)
```

# Alpha Diversity

Alpha diversity measures indicate species diversity within individual samples.

- **Richness:** The total number of unique species (or operational taxonomic units) present in a sample, without considering their abundance.
- **Shannon Index:** A measure of species diversity that accounts for both richness and evenness, giving higher values to communities with more evenly distributed species.
- **Fisher’s Alpha:** A parameter from the Fisher log-series model that estimates diversity by describing the relationship between species richness and abundance, particularly useful for highly diverse communities.
- **Pielou’s Evenness:** A measure of how evenly individuals are distributed among species in a community, ranging from 0 (uneven) to 1 (perfectly even).
- **Simpson Index:** A measure of dominance that quantifies the probability that two randomly selected individuals belong to the same species, with lower values indicating higher diversity.

## Hypothesis Setting

Group samples by flow rate.

```{r alpha_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Flow"

meta_table$Groups <- as.factor(as.character(paste("Flow: ", meta_table$sweetening.flow)))

meta_table$Type <- meta_table$Groups
meta_table$Type2 <- meta_table$Groups

meta_table$Connections <- NULL
```

```{r collate, include=FALSE}
#Adjust abund_table to contain only those rows that got selected in the Hypothesis space
abund_table<-abund_table[rownames(meta_table),]
#After adjustment, get rid of OTUs that are all empty
abund_table<-abund_table[,colSums(abund_table)>0]
#Adjust OTU taxonomy
taxa_table<-taxa_table[colnames(abund_table),]

#### Collate Taxonomy ####

abund_table <- collate_taxonomy(abund_table, taxa_table, which_level)
```

```{r alpha_diversity, include=FALSE}
#### Analysis ####

dt <- alpha_diversity(abund_table, taxa_table, meta_table, taxa_tree)

# Some absolutely fucking insufferable bug causes this to fail if its run in a package
# so fuck it I'm sticking it here instead
pval<-data.table::as.data.table(dt)[, list(pvalue = sprintf("%.2g", tryCatch({
   summary(stats::aov(value ~ .SD$.group.))[[1]][["Pr(>F)"]][1]
  }, error = function(e) NULL))),
  by = list(measure)
]
  
result <- alpha_diversity_2(dt, pval, meta_table, grouping_column)

df <- result[[1]]
df_pw <- result[[2]]

rm(result)
```

## Plots

```{r alpha_plot, echo=FALSE}
point_size = 5
point_opacity = 0.8
number_of_rows = 1
use_provided_colors = FALSE
pairwise_text_size = 7
exclude_pvalues_text_from_drawing = FALSE

plot <- plot_alpha_diversity(df, df_pw)
plot <- plot_theme_default(plot)

print(plot)
```

# Beta Diversity

Beta diversity measures indicate differences in species composition between samples, capturing community variation.

- **Bray-Curtis:** A dissimilarity index that quantifies differences between communities based on species abundance, ignoring shared absences.
- **Unweighted UniFrac:** A phylogenetic distance metric that compares communities based on the presence or absence of taxa, incorporating evolutionary relationships.
- **Weighted UniFrac:** Similar to unweighted UniFrac but also considers the relative abundance of taxa, giving more weight to dominant species in a community.

## Hypothesis Setting

Group samples by run and compare flow conditions.

```{r beta_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Flow"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

meta_table$Type <- as.factor(as.character(paste("Flow: ", meta_table$sweetening.flow)))
meta_table$Type2 <- meta_table$Type

meta_table$Connections <- NULL
```

```{r ref.label='collate', include=FALSE}
```

```{r beta_diversity, include=FALSE}
#### Analysis ####
kind <- "se" #sd se (sd is for drawing ellipse based on sd, se is for drawing ellipse based on standard errors)
vars <- c("run", "week", "BED", "wet.vs.dry", "sweetening.flow")

result <- beta_diversity(abund_table, taxa_table, meta_table, taxa_tree, PERMANOVA_variables = vars)

PCOA <- result[[1]]
PCOA_lines <- result[[2]]
df_ell <- result[[3]]
sol <- result[[4]]

#rm(result)
```

## Plots

```{r beta_plot, echo=FALSE}
point_size = 5
point_opacity = 0.8
# number_of_rows = 1
# use_provided_colors = FALSE

draw_glow=FALSE
draw_mean_values_text=FALSE
draw_confidence_intervals=TRUE
draw_ellipses_and_not_polygons=FALSE

opacity_ellipses_polygons=0.2
linesize_ellipses_polygons=0.8
linetype_ellipses_polygons="solid" #blank solid dashed dotted dotdash longdash twodash

exclude_legends=FALSE


plot <- beta_diversity_plot(PCOA, PCOA_lines, df_ell, sol, vars)

print(plot)
```

# Core Microbiome

The core microbiome captures the set of microbial taxa consistently found across a given group of samples.

## Abundance

```{r core_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Abstraction_Method"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

meta_table$Type <- NULL
meta_table$Type2 <- NULL
meta_table$Connections <- NULL
```

```{r ref.label='collate', include=FALSE}
```

```{r core_microbiome, echo=FALSE}
# Loop through and create a core plot for each unique group
plots <- list()
for (group in unique(meta_table$Groups)) {
  meta_sub <- meta_table[meta_table$Groups == group,]
  
  data <- core_microbiome(abund_table, meta_sub, taxa_table, taxa_tree, what_detection = "absolute", short_names = TRUE)
  
  plots <- append(plots, list(data))
}

print(do.call(gridExtra::grid.arrange, plots))
```

## Taxa Bars

```{r core_taxa, echo=FALSE}
data <- core_taxa(abund_table, meta_table, taxa_table, taxa_tree)
plot <- plot_core_taxa(data)

print(plot)
```

## Abundance-Occupancy Models

## Hypothesis Setting

## Plots

# Differential Expression

Differential expression analyses changes in abundance of taxa between conditions.

```{r de_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Abstraction_Method"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

meta_table$Type <- NULL
meta_table$Type2 <- NULL
meta_table$Connections <- NULL
```

```{r ref.label='collate', include=FALSE}
```

```{r de_analysis, echo=FALSE}
# Loop through and create a core plot for each unique group
plots <- list()
for (group in unique(meta_table$Groups)) {
  meta_sub <- meta_table[meta_table$Groups == group,]
  
  data <- core_microbiome(abund_table, meta_sub, taxa_table, taxa_tree, what_detection = "absolute", short_names = TRUE)
  
  plots <- append(plots, list(data))
}

print(do.call(gridExtra::grid.arrange, plots))
```

## Plots

```{r de_plot, echo=FALSE}
data <- core_taxa(abund_table, meta_table, taxa_table, taxa_tree)
plot <- plot_core_taxa(data)

print(plot)
```

# Ecological Community Assembly

Ecological communities are shaped by a blend of deterministic and stochastic processes. Deterministic processes, often referred to as niche-based, arise when differences in species’ traits or interactions determine which species can thrive in a particular environment. For example, factors like temperature or nutrient availability may favour some species over others, while competition, predation, or mutualistic partnerships can further refine which organisms succeed. Collectively, these forces are described as ‘selection’, reflecting how environmental filters and species traits work together to define community composition.

By contrast, stochastic processes, often referred to as neutral, operate more like chance events. In this model, all species are considered functionally equivalent, meaning community assembly is driven entirely by random births, deaths, dispersal, extinction, and speciation. This ‘ecological drift’ can be particularly influential when populations are small or selection is weak, leading to sometimes unpredictable shifts in which species persist.

Real communities seldom fall neatly into purely deterministic or purely stochastic categories. Instead, they emerge through a dynamic interplay of these forces, alongside diversification (the generation of new genetic variation) and dispersal (movement and establishment across locations). Another important concept is historical contingency, whereby past events create legacies that shape present communities. A key example is the ‘priority effect’, in which early-arriving species either hinder or help newcomers through resource use or environmental modifications. 

Understanding how all these factors interact remains a central challenge in ecology. In particular, wildly different processes can produce similar patterns, making interpretation of results challenging.

Measures of community assembly processes capture the ecological mechanisms which shape community composition.

- NRI
- NTI
- NST
- QPE

## Net Relatedness Index (NRI)

## Nearest Taxon Index (NTI)

## Normalised Stochasticity Ratio (NST)

```{r nst_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="By_Run"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))
```

```{r ref.label='collate', include=FALSE}
```

```{r nst_analysis, echo=FALSE}
df <- nst(abund_table, meta_table, taxa_table, taxa_tree)

p <- plot_nst(df)

print(p)
```

## Quantitative Process Estimate (QPE)

# Subset Regression

Used to select the set of environmental predictor variables which most efficiently model a given diversity metric.

## Hypothesis Setting

## Plots

# Variable Selection

Used to select two groups of taxa whose balance is most efficiently associated with an environmental variable.

## Hypothesis Setting

## Plots
