---
title: "Full Workflow"
date: "`r Sys.Date()`"
runtime: shiny
output: 
  html_document:
    fig_width: 16
    fig_height: 10
    fig_caption: true
    number_sections: true
    toc: true
    toc_float:
      collapsed: false
---

```{r setup, include=FALSE}
rm(list=ls())

knitr::opts_chunk$set(echo = TRUE)

# For testing
devtools::load_all("..")

# For production
# library(EcologyCore)
```

```{r params, include=FALSE}
#### Parameters ####

biom_path="/Users/reuben/test_data/data/Batch 1/asv_even_taxon.biom"
meta_path="/Users/reuben/test_data/data/metadata.csv"
tree_path="/Users/reuben/test_data/data/Batch 1/tree_rooted.nwk"
grouping_column<-"Groups"

which_level="Otus" #Otus Genus Family Order Class Phylum
taxa_rank <- which_level
```

# Introduction

This workflow runs through a number of ecological analyses in series as follows:

- **Alpha Diversity:** Richness, Shannon Index, Simpson Index, Fisher's Alpha, and Pielou's Evenness
- **Beta Diversity:** Bray-Curtis, Unweighted UniFrac, and Weighted UniFrac
- **Core Microbiome:** taxa bars and core heatmaps combine to show the most abundant species in each sample + the most prevalent in each group
- **Differential Expression:** DESeq2; shows significant differences in abundance between pairs of groups
- **Community Assembly Processes:** Normalised Stochasticity Ratio, Quantitative Process Estimate
- **Environmental Filtering:** NRI, NTI; environmental impact on evolution of phylogeny
- **Subset Regression:** regress metadata against diversity metrics (alpha, NRI, NTI, LCBD, etc.)
- **Variable Selection:** Selbal, CLR LASSO, CODA LASSO; selects two groups of taxa whose balance is most associated with the response (metadata) variable

In combination, these analyses are intended to give a complete picture of an ecological system.

```{r alpha_diversity, include=FALSE, child='alpha_diversity.Rmd'}

```

```{r beta_diversity, include=FALSE, child='beta_diversity.Rmd'}

```

```{r core_microbiome, include=FALSE, child='core_microbiome.Rmd'}

```

# Core Microbiome

The core microbiome captures the set of microbial taxa consistently found across a given group of samples.

## Abundance

```{r core_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Abstraction_Method"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

meta_table$Type <- NULL
meta_table$Type2 <- NULL
meta_table$Connections <- NULL
```

```{r ref.label='collate', include=FALSE}
```

```{r core_microbiome, echo=FALSE}
# Loop through and create a core plot for each unique group
plots <- list()
for (group in unique(meta_table$Groups)) {
  meta_sub <- meta_table[meta_table$Groups == group,]
  
  data <- core_microbiome(abund_table, meta_sub, taxa_table, taxa_tree, what_detection = "absolute", short_names = TRUE)
  
  plots <- append(plots, list(data))
}

print(do.call(gridExtra::grid.arrange, plots))
```

## Taxa Bars

```{r core_taxa, echo=FALSE}
data <- core_taxa(abund_table, meta_table, taxa_table, taxa_tree)
plot <- plot_core_taxa(data)

print(plot)
```

## Abundance-Occupancy Models

## Hypothesis Setting

## Plots

```{r differential_expression, include=FALSE, child='differential_expression.Rmd'}

```

# Differential Expression

Differential expression analyses changes in abundance of taxa between conditions.

```{r de_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="Abstraction_Method"

meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))

# control level should be the first level, set using relevel:
# meta_table$Groups <- relevel(meta_table$Groups, ref = "Run 1")

meta_table$Type <- NULL
meta_table$Type2 <- NULL
meta_table$Connections <- NULL
```

```{r ref.label='collate', include=FALSE}
```

Particularly relevant to environmental samples, note that the count matrix [should not be normalised](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts) prior to differential expression analysis. DESeq2 expects raw, un-normalised counts as input and accounts internally for differences in sequencing depth and coverage through its size factor estimation.

Therefore, if you suspect that some variable has systematically impacted total cell counts, such as differing sample volumes, you should include those variables as covariates in your design formula.

Note: In order to benefit from the default settings of the package, you should put the variable of interest at the end of the formula and make sure the control level is the first level.

```{r de_analysis, echo=FALSE}

result <- de_create(abund_table, meta_table, taxa_table, taxa_tree)

res_names <- result[[1]]
dds <- result[[2]]

rm(result)
```

## Results

```{r de_results, echo=FALSE}
# Loop through each comparison
selectInput("res_names", label = "Comparison:", choices = res_names, selected = res_names[1])

renderText({ res_names })
renderText({ de_summarise_results(dds, input$res_names) })
renderPlot({
  res <- de_get_results(dds, input$res_names)
  DESeq2::plotMA(res)
})
```

## Plots

```{r de_plot, echo=FALSE}

```

```{r community_assembly, include=FALSE, child='community_assembly.Rmd'}

```

# Ecological Community Assembly

Ecological communities are shaped by a blend of deterministic and stochastic processes. Deterministic processes, often referred to as niche-based, arise when differences in species’ traits or interactions determine which species can thrive in a particular environment. For example, factors like temperature or nutrient availability may favour some species over others, while competition, predation, or mutualistic partnerships can further refine which organisms succeed. Collectively, these forces are described as ‘selection’, reflecting how environmental filters and species traits work together to define community composition.

By contrast, stochastic processes, often referred to as neutral, operate more like chance events. In this model, all species are considered functionally equivalent, meaning community assembly is driven entirely by random births, deaths, dispersal, extinction, and speciation. This ‘ecological drift’ can be particularly influential when populations are small or selection is weak, leading to sometimes unpredictable shifts in which species persist.

Real communities seldom fall neatly into purely deterministic or purely stochastic categories. Instead, they emerge through a dynamic interplay of these forces, alongside diversification (the generation of new genetic variation) and dispersal (movement and establishment across locations). Another important concept is historical contingency, whereby past events create legacies that shape present communities. A key example is the ‘priority effect’, in which early-arriving species either hinder or help newcomers through resource use or environmental modifications. 

Understanding how all these factors interact remains a central challenge in ecology. In particular, wildly different processes can produce similar patterns, making interpretation of results challenging.

Measures of community assembly processes capture the ecological mechanisms which shape community composition.

- NRI
- NTI
- NST
- QPE

## Net Relatedness Index (NRI)

```{r nri_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
#label="By_Run"

#meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))
```

```{r ref.label='collate', include=FALSE}
```

```{r nri_analysis, echo=FALSE}
#df <- environmental_filtering(abund_table, meta_table, taxa_table, taxa_tree)

#p <- plot_environmental_filtering(df)

#print(p)
```

## Nearest Taxon Index (NTI)

## Normalised Stochasticity Ratio (NST)

```{r nst_hypothesis, include=TRUE}
#### Hypothesis ####

# Hypothesis 1
label="By_Run"

#meta_table$Groups <- as.factor(as.character(paste("Run ", meta_table$run)))
```

```{r ref.label='collate', include=FALSE}
```

```{r nst_analysis, echo=FALSE}
#df <- nst(abund_table, meta_table, taxa_table, taxa_tree)

#p <- plot_nst(df)

#print(p)
```

## Quantitative Process Estimate (QPE)

# Subset Regression

Used to select the set of environmental predictor variables which most efficiently model a given diversity metric.

## Hypothesis Setting

## Plots

# Variable Selection

Used to select two groups of taxa whose balance is most efficiently associated with an environmental variable.

## Hypothesis Setting

## Plots
